File: backend/app/models/energy.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import List, Optional, Dict, Any
from enum import Enum

class DatasetType(str, Enum):
    CONSUMPTION_REALTIME = "consumption_realtime"
    PRODUCTION_REALTIME = "production_realtime"
    WIND_PRODUCTION = "wind_production"
    CONSUMPTION_FORECAST = "consumption_forecast"
    PRICE_FORECAST = "price_forecast"

class EnergyDataPoint(BaseModel):
    """Individual energy data point"""
    timestamp: datetime
    value: float
    unit: str = "MW"

class EnergyData(BaseModel):
    """Energy dataset container"""
    dataset_id: int
    name: str
    dataset_type: DatasetType
    data: List[EnergyDataPoint]
    last_updated: datetime
    metadata: Optional[Dict[str, Any]] = None

class DifferentialPoint(BaseModel):
    """Production vs consumption differential point"""
    timestamp: datetime
    production: float
    consumption: float
    differential: float
    status: str = Field(..., description="surplus, deficit, or balanced")
    percentage: float = Field(..., description="Differential as percentage of consumption")

class DifferentialAnalysis(BaseModel):
    """Complete differential analysis"""
    analysis_period: str
    data: List[DifferentialPoint]
    summary: Dict[str, float]
    generated_at: datetime

class PriceData(BaseModel):
    """Electricity price data"""
    timestamp: datetime
    price: float
    unit: str = "EUR/MWh"
    area: str = "FI"

class SystemStatus(BaseModel):
    """System status information"""
    api_status: str
    cache_status: str
    last_data_update: datetime
    active_datasets: List[str]
    uptime_seconds: int

File: frontend/src/types/energy.ts
// Type definitions for energy data

export interface EnergyDataPoint {
  timestamp: Date | string;
  value: number;
  unit: string;
}

export interface EnergyData {
  dataset_id: number;
  name: string;
  dataset_type: string;
  data: EnergyDataPoint[];
  last_updated: Date | string;
  metadata?: Record<string, any>;
}

export interface PriceData {
  timestamp: Date | string;
  price: number;
  unit: string;
  area: string;
}

export interface DifferentialPoint {
  timestamp: Date | string;
  production: number;
  consumption: number;
  differential: number;
  status: 'surplus' | 'deficit' | 'balanced';
  percentage: number;
}

export interface DifferentialAnalysis {
  analysis_period: string;
  data: DifferentialPoint[];
  summary: {
    total_data_points: number;
    surplus_hours: number;
    deficit_hours: number;
    balanced_hours: number;
    surplus_percentage: number;
    deficit_percentage: number;
    total_surplus_mwh: number;
    total_deficit_mwh: number;
    net_balance_mwh: number;
  };
  generated_at: Date | string;
}

export enum DatasetType {
  CONSUMPTION_REALTIME = 'consumption_realtime',
  PRODUCTION_REALTIME = 'production_realtime',
  WIND_PRODUCTION = 'wind_production',
  CONSUMPTION_FORECAST = 'consumption_forecast',
}

export interface SystemStatus {
  status: string;
  timestamp: Date | string;
  services: {
    api: boolean;
    redis: boolean;
    fingrid: boolean;
    entsoe: boolean;
  };
}
